qualtran/bloqs/arithmetic/addition.py

```diff
@@ -266,6 +269,122 @@ def _add_oop_large() -> OutOfPlaceAdder:
 )
 
 
+@frozen
+class SimpleAddConstant(Bloq):
+    r"""Takes |x> to |x + k> for a classical integer `k`.
+
+    Applies addition to input register `|x>` given classical integer 'k'.
+
+    This is the simple version of constant addition because it involves simply converting the
+    classical integer into a quantum parameter and using quantum-quantum addition as opposed to
+    designing a bespoke circuit for constant addition based on the classical parameter.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+        k: The classical integer value to be added to x.
+        cvs: A tuple of control variable settings. Each entry specifies whether that
+            control line is a "positive" control (`cv[i]=1`) or a "negative" control `0`.
+        signed: A boolean condition which controls whether the x register holds a value represented
+            in 2's Complement or Unsigned. This affects the ability to add a negative constant.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+
+    References:
+        [Improved quantum circuits for elliptic curve discrete logarithms](https://arxiv.org/abs/2001.09580) Fig 2a
+    """
+
+    bitsize: int
+    k: int
+    cvs: Tuple[int, ...] = field(converter=lambda v: (v,) if isinstance(v, int) else tuple(v))
+    signed: bool
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        if len(self.cvs) > 0:
+            return Signature(
+                [Register('ctrl', bitsize=len(self.cvs)), Register('x', bitsize=self.bitsize)]
+            )
+        else:
+            return Signature([Register('x', bitsize=self.bitsize)])
+
+    def on_classical_vals(
+        self, x: 'ClassicalValT', **vars: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        if self.cvs:
+            ctrl = vars['ctrl']
+        else:
+            return {'x': x + self.k}
+
+        if np.equal(ctrl, self.cvs):
+            return {'ctrl': ctrl, 'x': x + self.k}
+        else:
+            return {'ctrl': ctrl, 'x': x}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, **regs: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+
+        # Assign registers to variables and allocate ancilla bits for classical integer k.
+        if len(self.cvs) > 0:
+            ctrls = regs['ctrl']
+        else:
+            ctrls = None
+        k = bb.allocate(n=self.bitsize)
+
+        # Get binary representation of k and split k into separate wires.
+        k_split = bb.split(k)
+        if self.signed:
+            binary_rep = list(iter_bits_twos_complement(self.k, self.bitsize))
+        else:
+            binary_rep = list(iter_bits(self.k, self.bitsize))
+        binary_rep.reverse()
```

Comment:
can you speak on this? This is an endianness thing, right? What endianness do you assume for the `x` register? We should make sure this is consistent throughout the library and we should document it somewhere

Reply:
Yeah, it's not really ideal, but it is because the functions return big-endian representations, but iterating through the qubits is indexed little-endian so I chose to reverse the output of the bit representation functions. I'll add a comment about this.